firestore.rules

```diff
@@ -216,18 +216,33 @@ service cloud.firestore {
       allow update: if (request.auth.uid == resource.data.creatorId || isAdmin())
                         && request.resource.data.diff(resource.data)
                         .affectedKeys()
-                        .hasOnly(['name', 'about', 'anyoneCanJoin', 'aboutPostId', 'pinnedItems','bannerUrl' ]);
+                        .hasOnly(['name', 'about', 'aboutPostId', 'pinnedItems','bannerUrl' ]);
       allow delete: if request.auth.uid == resource.data.creatorId;
 
       match /groupContracts/{contractId} {
-        allow write: if isGroupMember() || request.auth.uid == get(/databases/$(database)/documents/groups/$(groupId)).data.creatorId
+        allow write: 
+        if (
+        request.auth.uid == get(/databases/$(database)/documents/groups/$(groupId)).data.creatorId || 
+        isAdmin() || 
+        (
+        isGroupMember() && 
+        (
+        get(/databases/$(database)/documents/contracts/$(contractId)).data.creatorId == request.auth.uid || 
+        get(/databases/$(database)/documents/groups/$(groupId)/groupMembers/$(request.auth.uid)).data.role == 'admin' ||
+        get(/databases/$(database)/documents/groups/$(groupId)/groupMembers/$(request.auth.uid)).data.role == 'contributor'
```

Comment:
Doesn't matter that much, but I believe this will do the query twice. Instead, you could do something like
`['admin', 'contributor'].includes(get/ ......))`

Reply:
Might want to move this logic to a function too